src/model/transaction/removeRangeFromContentState.js

```diff
@@ -13,30 +13,266 @@
 
 'use strict';
 
+import type {BlockMap} from 'BlockMap';
 import type CharacterMetadata from 'CharacterMetadata';
 import type ContentState from 'ContentState';
 import type SelectionState from 'SelectionState';
-import type {List} from 'immutable';
 
-var Immutable = require('immutable');
+const ContentBlockNode = require('ContentBlockNode');
+const Immutable = require('immutable');
 
-function removeRangeFromContentState(
+const getNextDelimiterBlockKey = require('getNextDelimiterBlockKey');
+
+const {List, Map} = Immutable;
+
+const transformBlock = (
+  key: ?string,
+  blockMap: BlockMap,
+  func: (block: ContentBlockNode) => void,
+): void => {
+  if (!key) {
+    return;
+  }
+
+  const block = blockMap.get(key);
+
+  if (!block) {
+    return;
+  }
+
+  blockMap.set(key, func(block));
+};
+
+/**
+ * Ancestors needs to be preserved when there are non selected
+ * children to make sure we do not leave any orphans behind
+ */
+const getAncestorsKeys = (
+  blockKey: ?string,
+  blockMap: BlockMap,
+): Array<string> => {
+  const parents = [];
+
+  if (!blockKey) {
+    return parents;
+  }
+
+  let blockNode = blockMap.get(blockKey);
+  while (blockNode && blockNode.getParentKey()) {
+    const parentKey = blockNode.getParentKey();
+    parents.push(parentKey);
+    blockNode = blockMap.get(parentKey);
+  }
+
+  return parents;
+};
+
+/**
+ * Get all next delimiter keys until we hit a root delimiter and return
+ * an array of key references
+ */
+const getNextDelimitersBlockKeys = (
+  block: ContentBlockNode,
+  blockMap: BlockMap,
+): Array<string> => {
+  const nextDelimiters = [];
+
+  if (!block) {
+    return nextDelimiters;
+  }
+
+  let nextDelimiter = getNextDelimiterBlockKey(block, blockMap);
+  while (nextDelimiter && blockMap.get(nextDelimiter)) {
+    const block = blockMap.get(nextDelimiter);
+    nextDelimiters.push(nextDelimiter);
+
+    // we do not need to keep checking all root node siblings, just the first occurance
+    nextDelimiter = block.getParentKey()
+      ? getNextDelimiterBlockKey(block, blockMap)
+      : null;
+  }
+
+  return nextDelimiters;
+};
+
+const getNextValidSibling = (
```

Comment:
I'm curious what 'valid' means - it looks like it means that the block is not part of the section being removed. I'll keep reading and see if I'm right.

Reply:
I should maybe rename this. `getNextValidSibling` is used to identify blocks that are still existing on the new modified blockmap. Since blocks are removed by range and we have special rules for retaining blocks if they have children we need to make sure they are still existent on the newly updated blockmap